### PR TITLE
cpumanager: rollback state if updateContainerCPUSet failed

### LINKED ISSUE #63018: Internal PreStartContainer hook failed: not enough cpus available to satisfy request
/kind bug
**What happened**:
Sometimes when an deployment creates new pods they get scheduled on a node successfully but the containers fail to start with:

```
Warning  Failed                 14m (x8 over 15m)    kubelet, ip-172-20-162-142.ec2.internal  Internal PreStartContainer hook failed: not enough cpus available to satisfy request
```

**What you expected to happen**:
The scheduler shouldn't put pods on nodes without enough resources.

**How to reproduce it (as minimally and precisely as possible)**:
I don't know how it got into this situation. Deleting the pods and have the RS recreate them works around this issue.

**Anything else we need to know?**:
We're using the CPUManager with static policy which might contribute to this problem.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.0-beta.2", GitCommit:"63dad40a0391b7af32c34fdbf41fa199c3b247ad", GitTreeState:"clean", BuildDate:"2018-03-07T20:42:18Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:44:10Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration: AWS
- OS (e.g. from /etc/os-release): coreos

/sig scheduling

### PR DESCRIPTION
**Pull request by choury**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #63018

If `updateContainerCPUSet`  failed, the container will start failed. We should rollback the state to avoid CPU leak.
**Special notes for your reviewer**:

**Release note**:

```release-note
cpumanager: rollback state if updateContainerCPUSet failed
```

### CODE REVIEW

**Comment by ipuustin** [comment]
Hmm, looking at static policy, if `updateContainerCPUSet()` fails during `AddContainer()`, and we remove the container from state by calling `RemoveContainer()`, the reconciliation loop will try call `AddContainer()` again for it soon afterwards. I'm not sure if this is what we want. The current behavior is that the reconciliation loop will just call `updateContainerCPUSet()` again for the container.

**Comment by ipuustin** [comment]
Ah sorry, I spoke too soon. Indeed the container will not start if `AddContainer()` called from `PreStartHook()` fails. I was thinking of the other case where it is called from the reconciliation. This patch makes sense to me.

**Comment by ConnorDoyle** [comment]
/ok-to-test

Thanks for adding this! Please add some tests.

**Comment by choury** [comment]
@ConnorDoyle

**Comment by rphillips** [comment]
lgtm

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 1]
```diff
@@ -175,18 +175,17 @@ func (m *manager) AddContainer(p *v1.Pod, c *v1.Container, containerID string) e
 		return err
 	}
 	cpus := m.state.GetCPUSetOrDefault(containerID)
-	m.Unlock()
+	defer m.Unlock()
```
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go]: The previous code avoided holding the lock while doing the I/O involved in updateContainerCPUSet (this involves a gRPC call to the CRI.) Could we release and re-acquire the lock if necessary before the RemoveContainer call?

**Review by ConnorDoyle — COMMENTED** [review]
Tests look fine to me, just a couple of small comments.

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 2]
```diff
@@ -175,18 +175,17 @@ func (m *manager) AddContainer(p *v1.Pod, c *v1.Container, containerID string) e
...
 
 	if !cpus.IsEmpty() {
 		err = m.updateContainerCPUSet(containerID, cpus)
 		if err != nil {
 			glog.Errorf("[cpumanager] AddContainer error: %v", err)
-			return err
+			m.policy.RemoveContainer(m.state, containerID)
```
[the remark this answers is not in the document]
Comment (choury) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: I will add some error message here

File: pkg/kubelet/cm/cpumanager/cpu_manager.go (hunk 1 above)
Comment (choury) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: will update it.

**Comment by choury** [comment]
@ConnorDoyle 
Any other idea?

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 3]
```diff
@@ -181,12 +181,16 @@ func (m *manager) AddContainer(p *v1.Pod, c *v1.Container, containerID string) e
...
+			m.Lock()
+			err := m.policy.RemoveContainer(m.state, containerID)
+			if err != nil {
+				glog.Errorf("[cpumanager] AddContainer rollback state error: %v", err)
+			}
+			m.Unlock()
 		}
-	} else {
-		glog.V(5).Infof("[cpumanager] update container resources is skipped due to cpu set is empty")
+		return err
```
[the remark this answers is not in the document]
Comment (choury) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: I didn't change it, just did some refactor with it. Please review the whole function

[the remark this answers is not in the document]
Comment (ConnorDoyle) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: Got it, sorry.

**Pull request merged**